driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +185,362 @@ public int getMaxWireVersion() {
+            if (skipReleasingServerResourcesOnClose) {
+                serverCursor = null;
             }
-            endOperationInProgress();
 
-            if (result.getResults().isEmpty()) {
-                callback.onResult(null, null);
+            if (getServerCursor() != null) {
+                executeWithConnection(c -> {
+                    releaseServerAndClientResources(c);
+                    releaseServerResources(c);
+                    c.release();
+                }, (r, t) -> {
+                    // guarantee that regardless of exceptions, `serverCursor` is null and client resources are released
+                    serverCursor = null;
+                    releaseClientResources();
+                });
             } else {
-                callback.onResult(result.getResults(), null);
+                releaseClientResources();
             }
         }
-    }
 
-    private CommandCursorResult<T> initFromCommandCursorDocument(
-            final ServerAddress serverAddress,
-            final String fieldNameContainingBatch,
-            final BsonDocument commandCursorDocument) {
-        CommandCursorResult<T> cursorResult = new CommandCursorResult<>(serverAddress, fieldNameContainingBatch, commandCursorDocument);
-        cursor.set(cursorResult.getServerCursor());
-        count.addAndGet(cursorResult.getResults().size());
-        postBatchResumeToken = cursorResult.getPostBatchResumeToken();
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", cursorResult.getResults().size(),
-                    cursorResult.getCursorId(), cursorResult.getServerAddress()));
+        void onCorruptedConnection(@Nullable final AsyncConnection corruptedConnection) {
+            // if `pinnedConnection` is corrupted, then we cannot kill `serverCursor` via such a connection
+            AsyncConnection localPinnedConnection = pinnedConnection;
+            if (localPinnedConnection != null) {
+                assertTrue(corruptedConnection == localPinnedConnection);
+                skipReleasingServerResourcesOnClose = true;
+            }
         }
-        return cursorResult;
-    }
 
-    private void logQueryResult(final CommandCursorResult<T> result) {
-        LOGGER.debug(format("Received batch of %d documents with cursorId %d from server %s", result.getResults().size(),
-                result.getCursorId(), result.getServerAddress()));
-    }
+        <R> void executeWithConnection(final Consumer<AsyncConnection> action, final SingleResultCallback<R> callback) {
+            assertTrue(state != State.IDLE);
+            if (pinnedConnection != null) {
+                executeWithConnection(assertNotNull(pinnedConnection).retain(), null, action, callback);
+            } else {
+                assertNotNull(connectionSource).getConnection((conn, t) -> executeWithConnection(conn, t, action, callback));
+            }
+        }
```

Comment:
I suspect, that the technical problem Jeff pointed out in https://github.com/rozza/mongo-java-driver/pull/399/files?show-viewed-files=true&file-filters%5B%5D=#r1358948111 is part of a larger problem with this method, whose signature and implementation are surprising (maybe there are good reasons behind, but I don't immediately see them):

1. `Consumer<AsyncConnection>` is a weird type: it represents a seemingly synchronous action that consumes an asynchronous connection. This tells us that the action either does not meaningfully use the `AsyncConnection` (it may call `AsyncConnection.hashCode`, but can't call `AsyncConnection.commandAsync`), or that it has a callback captured. And indeed, we can see on line 195 `resourceManager.executeWithConnection(connection -> getMore(connection, cursor, callback), callback)` that it captures `callback`.
2. Now that we know that  `Consumer<AsyncConnection> action` captures `callback`, having `callback` as the second parameter of the `executeWithConnection` method is another surprising thing. At best, it suggests that the code structure may be suboptimal. At worst, it opens opportunities for `callback` to be accidentally completed more than once without a writer realizing that.
3) The exception `t` on line 419 `assertNotNull(connectionSource).getConnection((conn, t) -> executeWithConnection(conn, t, action, callback))` is not handled directly in the callback passed to `getConnection`, as we usually do, but instead, is passed further to `executeWithConnection`. This makes code more obscure and is surprising (we don't usually handle exceptions in callback-based code this way). Why is the error handled here this way?

Reply:
Thanks for the feedback, I've updated to use `AsyncCallableWithConnection` which fixes and improves the code readibility. Turns out I didn't grok `AsyncCallbackSupplier` properly at all.

Reply:
I don't understand why, but there is no "resolve conversation" button for me on this or any other comments I left. Ross, if you can, could you please resolve this thread?

P.S. I wonder how to proceed with the rest of the review process. Maybe GitHub will eventually start behaving.